logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt

```diff
@@ -0,0 +1,78 @@
+import com.wire.kalium.logic.di.MapperProvider
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.suspending
+import com.wire.kalium.logic.wrapApiRequest
+import com.wire.kalium.logic.wrapStorageRequest
+import com.wire.kalium.network.api.message.MLSMessageApi
+import com.wire.kalium.persistence.dao.ConversationDAO
+import com.wire.kalium.persistence.dao.ConversationEntity
+import io.ktor.util.decodeBase64Bytes
+import kotlinx.coroutines.flow.first
+
+interface MLSConversationRepository {
+
+    suspend fun establishMLSGroup(conversation: Conversation): Either<CoreFailure, Unit>
+
+}
+
+class MLSConversationDataSource(
+    val keyPackageRepository: KeyPackageRepository,
+    val mlsClientProvider: MLSClientProvider,
+    val mlsMessageApi: MLSMessageApi,
+    val conversationDAO: ConversationDAO,
+    val idMapper: IdMapper = MapperProvider.idMapper()
+): MLSConversationRepository {
+
+    override suspend fun establishMLSGroup(conversation: Conversation): Either<CoreFailure, Unit> = suspending  {
+        val groupID: String = conversation.groupId?.let { it } ?: run { return@suspending Either.Left(CoreFailure.MissingClientRegistration) }
+
+        getConversationMembers(conversation.id).flatMap { members ->
+            establishMLSGroup(groupID, members)
+        }
+    }
+
+    private suspend fun establishMLSGroup(groupID: String, members: List<UserId>): Either<CoreFailure, Unit> = suspending {
+        keyPackageRepository.claimKeyPackages(members).flatMap { keyPackages ->
+            mlsClientProvider.getMLSClient().flatMap { client ->
+                val clientKeyPackageList = keyPackages
+                    .map {
+                        Pair(
+                            CryptoQualifiedClientId(it.clientID, CryptoQualifiedID(it.userId, it.domain)),
+                            it.keyPackage.decodeBase64Bytes()
+                        )
+                    }
+
+                // TODO: send handshake when API is available
+                client.createConversation(groupID, clientKeyPackageList)?.let { (handshake, welcome) ->
+                    wrapApiRequest {
+                        mlsMessageApi.sendWelcomeMessage(MLSMessageApi.WelcomeMessage(welcome))
+                    }.flatMap {
+                        wrapStorageRequest {
+                            conversationDAO.updateConversationGroupState(ConversationEntity.GroupState.ESTABLISHED, groupID)
+                        }
+                    }.flatMap {
+                        Either.Right(Unit)
+                    }
+                } ?: run {
+                    Either.Right(Unit)
+                }
+            }
+        }
+    }
+
+    private suspend fun getConversationMembers(conversationId: ConversationId): Either<StorageFailure, List<UserId>> = wrapStorageRequest {
+        conversationDAO.getAllMembers(idMapper.toDaoModel(conversationId)).first().map { idMapper.fromDaoModel(it.user) }
+    }
+
+}
```

Comment:
not an kalium expert but I kind of feel that we got some buisness logic here that could be extracted to a use case, that way we avoid reimplementing things like getConversationMembers, that piece of code could live inside ConversationRepository for example,

Just a suggestion as I am getting more familiar with Kalium

Reply:
I can't move this code to the `ConversationRepository` and call it from here since then I would create a circular dependency. IMO I would leave like this for now and refactor the conversation repositories when we actually have some more substantial code duplication.

Reply:
I was actually thinking of combing user repo and this repo inside a use case when we move this code but your call on this one :)

Reply:
Just out of curiousity but where do you see that risk?

Reply:
In that if I split creating a group into two steps which are connected in the CreateGroupUseCase, there's a risk that the next time someone needs to create a group in another use case he/she might forget to do the second step.

Reply:
He could reuse that use case if he needs it in another use case :p i think I am used to putting more stuff into the use cases than repos :p anyway i got your point if you think this stuff belongs to repo ill Mark it resolved